### PR TITLE
External link includes selected text

### PR DESCRIPTION
**Pull request by tonyyates**
As described in issue #4328 selected text wasn't populated when adding
a external link. This commit fixes that issue by borrowing from the following functions :

getSelectedBlocksMap()
getSelectedBlocksList()
getSelectedText()
from <https://github.com/jpuri/draftjs-utils/blob/e81c0ae19c3b0fdef7e0c1b70d924398956be126/js/block.js#L106>

I am not sure new tests are needed?
Tested on Firefox, Chrome, Opera and Safari

New behaviour:

kapture 2018-03-14 at 16 12 47

### CODE REVIEW

File: client/src/components/Draftail/sources/ModalWorkflowSource.js [hunk 1]
```diff
@@ -35,7 +84,7 @@ export const getChooserConfig = (entityType, entity) => {
       // This does not initialise the modal with the currently selected text.
       // This will need to be implemented in the future.
```
Comment (gasman) [inline client/src/components/Draftail/sources/ModalWorkflowSource.js]: Please can you remove this comment, now that it's no longer true? :-)

**Comment by AlexsMathilda** [comment]
I've tested the external link includes selected text on Windows IE11, MS Edge, latest Chrome and Firefox. MacOS 10.13 Safari, latest Chrome and Firefox Android Chrome 7, iOS Safari 10.

the issue is fixed

**Comment by thibaudcolas** [comment]
👏 I'll have a look at this hopefully before the end of the sprint, otherwise later on, and take care of the tests at the same time.

File: client/src/components/Draftail/sources/ModalWorkflowSource.js (hunk 1 above)
Comment (tonyyates) [reply client/src/components/Draftail/sources/ModalWorkflowSource.js]: hmm I did start the commit, realised this and ganged the file before saving, I thought I had done this, will sort though. Thanks for reviewing 👍

**Comment by tonyyates** [comment]
@thibaudcolas I am happy to write the tests tomorrow if you can give me a few pointers, I am not sure how the client tests currently work but I am a keen student.

**Comment by thibaudcolas** [comment]
@tonyyates cool, let's do that tomorrow then 🙂

**Comment by tonyyates** [comment]
@gasman with thanks to @thibaudcolas for his help showing me how the tests work with jest. I have now included a number of tests that make this more robust. 

screenshot 2018-03-15 16 22 50

**Comment by thibaudcolas** [comment]
👏 sorry I didn't get to this during the sprint, I'll review + merge the PR this week.

File: client/src/components/Draftail/DraftUtils.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      start = end;
+      end = temp;
+    }
+
+    let selectedText = '';
+    for (let i = 0; i < selectedBlocks.size; i += 1) {
+      const blockStart = i === 0 ? start : 0;
+      const blockEnd = i === (selectedBlocks.size - 1) ? end : selectedBlocks.get(i).getText().length;
+      selectedText += selectedBlocks.get(i).getText().slice(blockStart, blockEnd);
+    }
```
Comment (thibaudcolas) [inline client/src/components/Draftail/DraftUtils.js]: I removed a `if selectedBlocks.size > 0` conditional for this logic – there doesn't seem to be any cases where there is 0 selected blocks.

File: client/src/components/Draftail/DraftUtils.test.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+          },
+        ],
+      });
+      let editorState = EditorState.createWithContent(content);
+
+      let selection = editorState.getSelection();
+      selection = selection.merge({
+        anchorOffset: 8,
+        focusOffset: 4,
+        isBackward: true,
```
Comment (thibaudcolas) [inline client/src/components/Draftail/DraftUtils.test.js]: Boolean

File: client/src/components/Draftail/sources/ModalWorkflowSource.test.js [hunk 4]
```diff
@@ -2,13 +2,15 @@ import React from 'react';
...
 import ModalWorkflowSource, { getChooserConfig, filterEntityData } from './ModalWorkflowSource';
+import * as DraftUtils from '../DraftUtils';
 import { EditorState, convertFromRaw, AtomicBlockUtils, RichUtils, Modifier } from 'draft-js';
 
 global.ModalWorkflow = () => {};
 
 describe('ModalWorkflowSource', () => {
   beforeEach(() => {
     jest.spyOn(global, 'ModalWorkflow');
+    jest.spyOn(DraftUtils, 'getSelectionText').mockImplementation(() => '');
```
Comment (thibaudcolas) [inline client/src/components/Draftail/sources/ModalWorkflowSource.test.js]: Added mocking for this for the related test suites.

**Comment by tomdyson** [comment]
Thanks @tonyyates and @thibaudcolas!

**Comment by tonyyates** [comment]
A pleasure to get involved :-)

Thanks for everything you have done to help @thibaudcolas

**Pull request merged**